libs/src/services/solanaWeb3Manager/rewards.js

```diff
@@ -0,0 +1,767 @@
+  data: instruction.data,
+  keys: instruction.keys.map(({ isSigner, pubkey, isWritable }) => ({
+    pubkey: pubkey.toString(),
+    isSigner,
+    isWritable,
+  })),
+})
+
+/**
+ * Derives the 'transfer account' - the account which represents a single successful disbursement
+ * and is used to dedupe - from the transferId and other info
+ *
+ * @param {string} transferId
+ * @param {PublicKey} rewardProgramId
+ * @param {PublicKey} rewardManager
+ * @returns {Promise<PublicKey>}
+ */
+const deriveTransferAccount = async (
+  transferId,
+  rewardProgramId,
+  rewardManager
+) => {
+  const seed = Uint8Array.from([
+    ...encoder.encode(TRANSFER_PREFIX),
+    ...encoder.encode(transferId),
+  ])
+  const [_, derivedAddress] = await findProgramAddressWithAuthority(
+    rewardProgramId,
+    rewardManager,
+    seed
+  )
+  return derivedAddress
+}
+
+/**
+ * Derives the account to store messages for a single challenge
+ *
+ * @param {string} transferId
+ * @param {PublicKey} rewardsProgramId
+ * @param {PublicKey} rewardManager
+ * @returns {Promise<[PublicKey, PublicKey, number]>}
+ */
+const deriveMessageAccount = async (
+  transferId,
+  rewardsProgramId,
+  rewardManager
+) => {
+  const encodedPrefix = encoder.encode(VERIFY_TRANSFER_SEED_PREFIX)
+  const encodedTransferId = encoder.encode(transferId)
+  const seeds = Uint8Array.from([...encodedPrefix, ...encodedTransferId])
+  return findProgramAddressWithAuthority(rewardsProgramId, rewardManager, seeds)
+}
+
+/**
+ * Derives a program address from a program ID and pubkey as seed.
+ * Returns the new pubkey and bump seeds.
+ *
+ * @param {PublicKey} programId
+ * @param {PublicKey} pubkey
+ * @returns {Promise<[PublicKey, number]>}
+ */
+const findProgramAddressFromPubkey = (programId, pubkey) => {
+  return PublicKey.findProgramAddress(
+    [pubkey.toBytes().slice(0, 32)],
+    programId
+  )
+}
+
+/**
+ * Finds a 'derived' address by finding a programAddress with
+ * seeds array  as first 32 bytes of base + seeds
+ * Returns [derivedAddress, bumpSeed]
+ *
+ * @param {PublicKey} programId
+ * @param {PublicKey} base
+ * @param {Uint8Array} seed
+ * @returns {Promise<[PublicKey, number]>}
+ */
+const findProgramAddressFromPubkeyAndSeeds = (programId, base, seed) => {
+  return PublicKey.findProgramAddress(
+    [base.toBytes().slice(0, 32), seed],
+    programId
+  )
+}
+
+/**
+ * Finds a program address, using both seeds, pubkey, and the derived rewards manager authority.
+ * Return [rewardManagerAutuhority, derivedAddress, and bumpSeeds]
+ *
+ * @param {PublicKey} programId
+ * @param {PublicKey} rewardManager
+ * @param {Uint8Array} seed
+ * @returns {Promise<[PublicKey, PublicKey, number]>}
+ */
+const findProgramAddressWithAuthority = async (
+  programId,
+  rewardManager,
+  seed
+) => {
+  // Finds the rewardManagerAuthority account by generating
+  // a PDA with the rewardsMnager as a seed
+  const [rewardManagerAuthority] = await findProgramAddressFromPubkey(
```

Comment:
not sure if these are expensive, but maybe worth a Promise.all?

Reply:
would be a pretty ugly destructuring with a promise all imo - these are pretty quick afaict, don't think it's worth it